### ISSUE TITLE
[naming-convention] camelCase and strictCamelCase don't allow for multiple capital letter humps

### ISSUE THREAD

**Issue by IAMtheIAM** [typescript-eslint/typescript-eslint#2073]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/ "@typescript-eslint/naming-convention": [
            "error",
            {
                "selector": "default",
                "format": ["strictCamelCase"]
            },
    ]
  }
}
```

```TS
const myCamelCaseConst = 'foot' ;
```

**Expected Result**
No error
**Actual Result**
  `112:13  error  Variable name myCamelCase must match one of the following formats: strictCamelCase  @typescript-eslint/naming-convention`

**Additional Info**
The naming convention for camelCase and stringCamelCase should allow for multiple capital letters to occur in the name, because camels can have multiple humps.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.0.0` |
| `@typescript-eslint/parser`        | `3.0.0` |
| `TypeScript`                       | `3.7.4` |
| `ESLint`                           | `7.0.0` |
| `node`                             | `12.16.3` |
| `npm`                              | `6.14.4` |

**Comment by bradzacher** [comment]
I'm unable to repro this against master.
Could you please provide a repro repo?

**Comment by IAMtheIAM** [comment]
I will make one soon and update.

**Comment by IAMtheIAM** [comment]
I was able to resolve by using a regex that targets camelCaseVars with multiple capitals, because the default name didn't work. In case anyone wants that regex:
```
                    {
                        "selector": "variable",
                        "types": ["boolean"],
                        "format": [], // required, even as an empty array if using custom regex
                        "custom": {
                            "regex": "([a-z][a-z0-9]+)?((\d)|([A-Z0-9][a-z0-9]+))*([A-Z])?", // this check for a camelCaseConst that allows multiple humps i.e. capitals in the middle, but not consecutively.
                            "match": true
                        }
                    },
```

**Comment by bradzacher** [comment]
Ah - that is actually working as expected.

This is explained in the docs here:
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/naming-convention.md#format-options

When you define a format prefix, that prefix is "trimmed" from the name before evaluating the next step.

So in your case your config is:
```json
{
  "selector": "variable",
  "types": ["boolean"],
  "format": ["strictCamelCase"],
  "prefix": ["is", "should", "has", "can", "did", "will"]
}
```

So the rule does:

1) `name = hasUnchecked`
1) validate leading underscore - none = pass
1) validate trailing underscore - none = pass
1) validate prefix - has correct prefix = pass
  - Trim prefix - `name = Unchecked`
1) validate suffix - none = pass
1) validate custom - none = pass
1) validate format - `name === Unchecked`, so it doesn't match `strictCamelCase` = fail

The configuration works this way on purpose so that you can do things like require a prefix `T` for all generic parameters, but also require `StrictPascalCase`. If we didn't trim the prefix, then `TFoo` wouldn't match the strict format.

So for your use case, if you instead use `StrictPascalCase` as your format, this will work as expected.

**Comment by IAMtheIAM** [comment]
Ohh you know I did stumble upon that discovery just before you mentioned it here! But thanks for explaining that because I missed that caveat originally.

I was able to get all my checks to pass using regex but I'll see about migrating them over to your presets since they are much clearer.

I did notice that I wasn't able to combine your preset "StrictCamelCase" with my own custom regex... is that intentional? It would be good to use the presets, and also add a custom regex option to it as well.

**Comment by bradzacher** [comment]
you should be able to combine both `custom` and `format`. The rule checks the two in isolation - checking the `custom` first, then the `format`.
